Post/regression test (#30)
makinarocks/makinarocks.github.io: _posts/2021-02-10-Regresssion-Test.md

---
layout: post
title: Regression Test, Are you sure?
author: wontak ryu
categories: [test]
image: assets/images/2020-02-10-Performance-Test/13_.gif
---

## 들어가며

마키나락스는 AI Project를 넘어 AI Product로 나아가고 있습니다. Product로 나아가는 여정 속에서 재미있는 엔지니어링 이슈들이 생겼습니다.

> 특히, 마키나락스가 제공하는 Machine Learning Software의 성능 대해서 신속하고 정확한 검증에 대한 필요가 있었습니다. 마키나락스에서는 Unit Test는 기본적으로 적용하고 있지만, Unit Test만으로는 성능을 보장 할 수 없습니다.

Reviewer: ```suggestion
특히 마키나락스에서 제공하는 Machine Learning Software의 성능을 신속하고 정확하게 검증할 필요가 있었습니다. Product 개발에 Unit Test가 기본적으로 적용되고 있지만, 단순히 Unit Test만으로는 성능을 보장 할 수 없습니다.
```

- Machine Learning Software라는 표현이 모호한 것 같습니다. "AI Product에 내장된 Model" 또는 "AI Product의 Algorithm" 등으로 변경하는 건 어떨까요?

Author: 모델뿐만 아니라 데이터 파이프라인도 포함해야해서 조금 추상적으로 작성한 거 같습니다. 제 생각에는 "AI Product에 내장된 Model" 또는 "AI Product의 Algorithm"은 모델에 대한 것만 언급하는 것 같은데 어떻게 생각하시나요?

Author: machine learning sotfware에 대한 설명 추가하기

이번 포스트에서는 Machine Learning Software에 대한 성능검증을 어떤 방식으로 진행하고 있는지 공유드리도록 하겠습니다.

## Problem: Can't find the cause of the lower performance!

> 우선, 내부에서 겪었던 문제에 대해서 공유드리도록 하겠습니다. 마키나락스에서는 협업기반의 개발문화를 가지고 있습니다. 개발을 하면서, Branch들이 Pull-Request를 날리고 Review하고 Merge되는 과정을 반복하게 됩니다.

Reviewer: ```suggestion
우선 내부에서 AI Product를 개발하며 겪었던 문제에 대해서 공유드리도록 하겠습니다. 마키나락스에서는 GitHub를 통해 협업을 관리하고  있습니다. 따라서 개발자들은 작업이 어느 정도 완료되면 작업 중인 Branch에서 Pull-Request를 날리게 됩니다. 그럼 다른 개발자들이 해당 코드들을  Review하고 이상이 없으면 Merge 합니다.
```

- "협업 기반의 개발 문화"가 무엇인지 잘 모르겠습니다ㅠ
- Git의 워크플로우인 만큼 GitHub를 사용하고 있다는 것을 알려주면 좋을 것 같습니다.

Author: 네 반영하겠습니다.

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Performance-Test/1.png" alt="Gitflow Workflow" width="60%">
  <figcaption style="text-align: center;">[그림1] - Gitflow Workflow [16]</figcaption>
</p>
</figure>

> 내부에 Unit Test가 구현되어 있어서, 어느정도의 안전성이 검증된 상태였습니다. 하지만, 마키나락스는 Machine Learning Software를 다루는 조직이기 때문에 더 높은 수준의 Test System이 필요했습니다. 

Reviewer: "더 높은 수준의 Test System이 필요"한 이유에 대해서 언급해주면 어떨까요?

-> "Machine Learning의 특성상 안정성 뿐만 아니라 성능이 중요하다 이에 대한 검증이 필요하다."

Author: 반영했습니다.

변화량에 대해서 Debugging 비용을 그래프로 그려보면 [그래프1]이 나옵니다. 위에서 언급했듯이, 복합적인 원인에 의해서 성능저하 이슈가 발생했을 경우, 모든 변화의 조합에 대해서 실험 및 테스트를 진행해야합니다. (이 때, 모든 변화는 독립적이라는 가정을 하고 계산하였습니다.)

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Performance-Test/5.png" alt="Gitflow Workflow" width="60%">
  <figcaption style="text-align: center;">[그림5] - 경우의 수</figcaption>
</p>
</figure>
<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Performance-Test/graph1.png" alt="Gitflow Workflow" width="40%">
  <figcaption style="text-align: center;">[그래프1] - Cost of Debugging</figcaption>
</p>
</figure>

그럼 "언제 Regression Test를 진행해야하는가?" 라는 의문이 들 수 있습니다. [그래프1]을 근거로 내부에서는 "As Soon As Possible"(가능한 빨리)라는 결론을 내리게 되었고, 이를 위해서 쉽고 빠른 Regression Test Pipeline을 구성하였습니다.

## Trial and Errors

Regression Test Pipeline을 만들기 위해서, 여러가지 시행착오를 겪었습니다. 겪었던 시행착오를 통해서, 필요했던 **추상화 과정**에 대해서 설명드리겠습니다.

우선, 자동화 도구로 Jenkins를 활용하였습니다. [[1]](#ref-1) Jenkins는 소프트웨어 개발 시 지속적으로 통합 서비스를 제공하는 툴입니다. 비교적 높은 자유도가 있었고, 자동화 도구로 접근성이 좋다고 판단했습니다. 

### Pipeline #1: Dependent on Repository

> 첫 번째로 구현한 Pipeline은 아래 [그림6]에서 볼 수 있습니다. Jenkins Container가 Regression Test 대상이 되는 Repository의 Requiremnts를 미리 가지고 있습니다. 학습에 필요한 데이터의 경우 NAS에 저장해두고 요청 시 접근하여 사용합니다. GitHub에서 Test요청을 보내면, Regression Test를 진행하게 됩니다. 이런 구조는 한 Repository에 의존성을 가지게 된다는 문제를 가지고 있습니다.

Reviewer: ```suggestion
첫 번째로 구현한 Pipeline은 아래 [그림6]에서 볼 수 있습니다. Jenkins Container가 Regression Test 대상이 되는 Repository의 Requirements를 미리 가지고 있습니다. 학습에 필요한 데이터의 경우 NAS에 저장해두고 요청 시 접근하여 사용합니다. GitHub에서 Test요청을 보내면, Regression Test를 진행하게 됩니다. 

이런 구조는 한 Repository에 의존성을 가지게 된다는 문제를 가지고 있습니다.
```

- 문단을 나누어서 아래의 구조가 가지는 문제점을 보다 명확히 해주면 이해가 조금 더 잘 될 것 같습니다.

<figure class="image" style="align: center;">
<p align="center">

>   <img src="/assets/images/2020-02-10-Performance-Test/6.png" alt="Gitflow Workflow" width="60%">

Reviewer: 모든 이미지에 "Decktop"으로 표기되어 있는데, 맞는 표현인가요?

Author: 앗 이런... 오타입니다! 감사합니다.

Author: 반영했습니다.

  <figcaption style="text-align: center;">[그림6] - Pipeline #1</figcaption>
</p>
</figure>

### Pipeline #2: Independent on Repository, But Inefficient

두 번째로 구현한 Pipeline은 아래 [그림7]에서 볼 수 있습니다. Pipeline #1과 다르게 Jenkins Container가 Repository에 정의된 Dockerfile을 기반으로 Regression Test Container를 만듭니다. 이를 통해서 Repository에 의존성을 가지던 문제를 해결할 수 있었습니다. 하지만, Docker Image를 Build하는 작업은 상당히 오랜시간이 걸리기 때문에, 비효율적이라는 문제가 있었습니다.

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Performance-Test/7.png" alt="Gitflow Workflow" width="60%">
  <figcaption style="text-align: center;">[그림7] - Pipeline #2</figcaption>
</p>
</figure>

### Pipeline #3: InDependent on Repository, But!

> 첫 번째로 구현한 Pipeline은 아래 [그림7]에서 볼 수 있습니다. Docker Image는 Requirements가 변경되었을 때만 Update가 필요했습니다. 따라서, 미리 DockerImage를 만들어두고, Jenkins Container가 이를 받아서 사용하도록 변경하였습니다. Pipeline #3과 비교해봤을 때, 효율적이었습니다.

Reviewer: ```suggestion
첫 번째로 구현한 Pipeline은 아래 [그림7]에서 볼 수 있습니다. Docker Image는 Requirements가 변경되었을 때만 Update가 필요했습니다. 따라서 미리 DockerImage를 만들어 두고 Jenkins Container가 이를 받아서 사용하도록 구조를 변경하였습니다. Pipeline #3과 비교해봤을 때 효율적이었습니다.
```

- 아래 그림은 [그림 8] 입니다. 확인 부탁드리겠습니다!
- 마지막 문장 Pipeline #3과 비교했을 때 효율적이라는 표현이 나오는데 이 부분도 확인 부탁드리겠습니다!

Author: 반영했습니다. 감사합니다.

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Performance-Test/8.png" alt="Gitflow Workflow" width="60%">
  <figcaption style="text-align: center;">[그림8] - Pipeline #3</figcaption>
</p>
</figure>

### Device Dependency

> 하지만, Pipeline #1 ~ #3이 모두 공통적으로 한 컴퓨팅 자원에 의존적이라는 문제가 있었습니다. 예를 들어서, Regression Test에 사용하는 컴퓨팅자원에 만약 다른 작업이 돌아가고 있었다면, Regression Test가 아예 작동하지 못하거나, 다른 작업을 망칠 수도 있습니다. [그림9]를 보면, 3개의 노트북이 MRX-Desktop1에 접속하여 사용하고 있는 모습을 볼 수 있습니다. 만약 Jenkins Container가 MRX-Desktop1에서 작동하고 있다면, Regression Test가 정상적으로 작동하지 않을 것입니다.

Reviewer: ```suggestion
하지만 Pipeline #1 ~ #3은 모두 공통적으로 한 컴퓨팅 자원에 의존적이라는 문제가 있습니다. 예를 들어 Regression Test에 사용하는 컴퓨터에서 다른 작업이 돌아가고 있다면 Regression Test가 아예 작동하지 못하거나, 다른 작업을 망칠 수도 있습니다. [그림9]를 보면, 3개의 노트북이 MRX-Desktop1에 접속하여 사용하고 있는 모습을 볼 수 있습니다. 만약 Jenkins Container가 MRX-Desktop1에서 작동하고 있다면, Regression Test가 정상적으로 작동하지 않을 것입니다.
```

- 다른 작업을 망치게 되는 이유가 무엇일까요? Out of Memory 이슈가 떠오르기는 하는데 추가해주시면 좋을 것 같습니다!

Author: 감사합니다.

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Performance-Test/9.png" alt="Gitflow Workflow" width="60%">
  <figcaption style="text-align: center;">[그림9] - Problem of Device Dependency</figcaption>
</p>
</figure>

> 또한, 다른 MRX-Desktop2, 3를 보면 컴퓨팅 자원이 여유있다는 것을 알 수 있습니다. 이런 자원을 효율적으로 사용하기 위해서, Device Dependency를 제거하는 작업이 필요했습니다.

Reviewer: ```suggestion
또한 다른 MRX-Desktop2, 3를 보면 컴퓨팅 자원이 여유있다는 것을 알 수 있습니다. 자원을 효율적으로 사용하기 위해서 Device Dependency를 제거하는 작업이 필요했습니다.
```

- Device Dependency가 무엇인지 설명을 추가해주면 좋을 것 같습니다.

### Pipeline #4: InDependent on Device

> Device Dendency를 해결하기 위해서, Kubernetes를 사용하였습니다. [[2]](#ref-2) 대략적으로 Kubernetes에 대해서 알고 싶으신 분들은 다음 Reference를 참고하시는 것을 추천드립니다. [[6]](#ref-2)

Reviewer: ```suggestion
Device Dendency를 해결하기 위해서 Kubernetes를 사용하였습니다[[2]](#ref-2). Kubernetes에 대해서 알고 싶으신 분들은 Reference를 참고하시는 것을 추천드립니다[[6]](#ref-2).
```

Author: 찾아보니 잘못된 것들이 꽤 많을듯합니다. 감사합니다.

  <figcaption style="text-align: center;">[그림10] - Kubernetes in Makinarocks </figcaption>
</p>
</figure>

Jenkins Container의 역할은 특정 Device내에서 Container로 Regression Test를 진행하는 것이 아닙니다. Jenkins Container는 미리 정의된 컴퓨팅 자원 스펙에 해당하는 Ray Cluster를 만드는 것입니다. [[3]](#ref-2) 여기서 Ray Cluster의 역할은 Regression Test를 병렬적으로 진행하기 위한 목적으로 사용되고, 작업이 끝나게 되면 Ray Cluster는 사라지게 됩니다. 참고로 [그림10]에서 구성한 Cluster와 Ray Cluster는 다른 역할을 합니다. [그림10]은 자원자체를 묶는 작업을 의미한다면, Ray Cluster는 이미 묶인 자원을 활용하는 것입니다. 

**이번 포스팅에서 Kubernetes와 Ray Cluster에 대해서 자세히 다루지는 않겠지만, 수요가 있다면 마키나락스에서 Kubernetes를 활용하는 방법을 다룰 예정입니다. 혹시 관심있으신 분은 댓글 남겨주시면 감사하겠습니다.**

이제 Kubernetes 그리고 Ray Cluster를 활용하여, [그림11]과 같은 Pipeline을 구축하였습니다. Repository에 의존성을 제거하였으며, Docker Image도 미리 만들어둔 Image를 활용하였습니다. 또한, Device에 대한 의존성을 제거하여, 내부의 컴퓨팅 자원을 더욱 효율적으로 사용할 수 있습니다.

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Performance-Test/11.png" alt="Gitflow Workflow" width="60%">
  <figcaption style="text-align: center;">[그림11] - Pipeline #4 </figcaption>
</p>
</figure>

## (Selected) Method: Self-Hosted Runner in GitHub Action

> 위에서, Pipeline #1 ~ #4까지 살펴볼 수 있었습니다. 하지만, Jenkins에 익숙하지 않다보니 기술적인 이슈가 발생할때 대처하는데 쉽지 않았습니다. 특히 Kubernetes환경에서 jenkins를 활용하기 위해서는 조금 더 많은 지식이 필요했습니다. 아쉽게도 이에 대한 문서를 쉽게 찾을 수 없어서, 유지보수 측면에서 아쉬움이 있었습니다.

Reviewer: ```suggestion
Pipeline #1부터 #4까지 모두 Jenkins를 사용하고 있습니다. 하지만 Jenkins라는 툴에 익숙하지 않다보니 기술적인 이슈가 발생했을 때 대처하는데 쉽지 않았습니다. 특히 Kubernetes 환경에서 jenkins를 활용하기 위해서는 조금 더 많은 지식이 필요했습니다. 아쉽게도 이에 대한 문서를 쉽게 찾을 수 없어서 유지보수 측면에서 아쉬움이 있었습니다.
```

- 조금 더 많은 지식이 무엇인지, 왜 필요한지 함께 써주면 좋을 것 같습니다.

---
layout: post
title: Regression Test, Are you sure?
author: wontak ryu
categories: [test]
image: assets/images/2020-02-10-Regression-Test/13_.gif
---

## 들어가며

마키나락스는 AI Project를 넘어 AI Product로 나아가고 있습니다. Product로 나아가는 여정 속에서 재미있는 엔지니어링 이슈들이 생겼습니다.
특히, 마키나락스가 제공하는 Machine Learning Software의 성능 대해서 신속하고 정확한 검증에 대한 필요가 있었습니다. 마키나락스에서는 Unit Test는 기본적으로 적용하고 있지만, Unit Test만으로는 성능을 보장 할 수 없습니다.

> 이번 포스트에서는 Machine Learning Software에 대한 성능검증을 어떤 방식으로 진행하고 있는지 공유드리도록 하겠습니다.

Reviewer 2: ```suggestion
이번 포스트에서는 계속에서 개발되는 Machine Learning Software에 대한 성능검증을 어떤 방식으로 진행하고 있는지 공유드리도록 하겠습니다.
```
코드가 계속 변화하는 중이라는 부분이 들어가면 좋을 것 같습니다

## Problem: Can't find the cause of the lower performance!

우선 내부에서 AI Product를 개발하며 겪었던 문제에 대해서 공유드리도록 하겠습니다. 마키나락스에서는 GitHub를 통해 협업을 관리하고 있습니다. 따라서 개발자들은 작업이 어느 정도 완료되면 작업 중인 Branch에서 Pull-Request를 날리게 됩니다. 그럼 다른 개발자들이 해당 코드들을  Review하고 이상이 없으면 Merge 합니다.

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/1.png" alt="Gitflow Workflow" width="60%">
  <figcaption style="text-align: center;">[그림1] - Gitflow Workflow [16]</figcaption>
</p>
</figure>

> 내부에 Unit Test가 구현되어 있어서, 어느정도의 안전성이 검증된 상태였습니다. 하지만, 마키나락스는 Machine Learning Software를 다루는 조직이기 때문에 더 높은 수준의 Test System이 필요했습니다. Machine Learning의 특성상 안정성 검증뿐만 아니라 성능검증도 함께 이루어져야합니다. 현재 Test system내에서 "Unit Test는 통과하지만, 성능저하를 일으키는 Commit"같은 경우에는 추적을 할 수 없다는 문제가 있습니다. 실제로 겪었던 사례가 있습니다. Activation Function을 선택할 때, Relu를 사용하면 특정 모델의 성능저하가 일어나는 것을 발견하였습니다. Relu가 잘못된 알고리즘은 아니나, 특정모델에서는 정상적으로 작동하지 않았습니다.

Reviewer 2: ```suggestion
내부에 Unit Test가 구현되어 있어서, 어느정도의 안전성이 검증된 상태였습니다. 하지만, 마키나락스는 Machine Learning Software를 다루는 조직이기 때문에 더 높은 수준의 Test System이 필요했습니다. Machine Learning의 특성상 안정성 검증뿐만 아니라 성능검증도 함께 이루어져야합니다. 현재 Test System내에서 "Unit Test는 통과하지만, 성능저하를 일으키는 Commit"같은 경우에는 추적을 할 수 없다는 문제가 있습니다. 실제로 겪었던 사례가 있습니다. Activation Function을 선택할 때, Relu를 사용하면 특정 모델의 성능저하가 일어나는 것을 발견하였습니다. Relu가 잘못된 알고리즘은 아니나, 특정모델에서는 정상적으로 작동하지 않았습니다.
```

변화량에 대해서 Debugging 비용을 그래프로 그려보면 [그래프1]이 나옵니다. 위에서 언급했듯이, 복합적인 원인에 의해서 성능저하 이슈가 발생했을 경우, 모든 변화의 조합에 대해서 실험 및 테스트를 진행해야합니다. (이 때, 모든 변화는 독립적이라는 가정을 하고 계산하였습니다.)

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/5.png" alt="Gitflow Workflow" width="60%">
  <figcaption style="text-align: center;">[그림5] - 경우의 수</figcaption>
</p>
</figure>
<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/graph1.png" alt="Gitflow Workflow" width="40%">
  <figcaption style="text-align: center;">[그래프1] - Cost of Debugging</figcaption>
</p>
</figure>

그럼 "언제 Regression Test를 진행해야하는가?" 라는 의문이 들 수 있습니다. [그래프1]을 근거로 내부에서는 "As Soon As Possible"(가능한 빨리)라는 결론을 내리게 되었고, 이를 위해서 쉽고 빠른 Regression Test Pipeline을 구성하였습니다.

## Trial and Errors

Regression Test Pipeline을 만들기 위해서, 여러가지 시행착오를 겪었습니다. 겪었던 시행착오를 통해서, 필요했던 **추상화 과정**에 대해서 설명드리겠습니다.

우선, 자동화 도구로 Jenkins를 활용하였습니다 [[1]](#ref-1). Jenkins는 소프트웨어 개발 시 지속적으로 통합 서비스를 제공하는 툴입니다. 비교적 높은 자유도가 있었고, 자동화 도구로 접근성이 좋다고 판단했습니다. 

### Pipeline #1: Dependent on Repository

> 첫 번째로 구현한 Pipeline은 아래 [그림6]에서 볼 수 있습니다. Jenkins Container가 Regression Test 대상이 되는 Repository의 Requirements를 미리 가지고 있습니다. 학습에 필요한 데이터의 경우 NAS에 저장해두고 요청 시 접근하여 사용합니다. GitHub에서 Test요청을 보내면, Regression Test를 진행하게 됩니다. 

Reviewer 2: ```suggestion
첫 번째로 구현한 Pipeline은 아래 [그림6]에서 볼 수 있습니다. Jenkins Container가 Regression Test 대상이 되는 Repository의 Requirements를 미리 가지고 있습니다. 학습에 필요한 데이터의 경우 원격 저장소에 저장해두고 요청 시 접근하여 사용합니다. GitHub에서 Test요청을 보내면, Regression Test를 진행하게 됩니다. 
```

  <figcaption style="text-align: center;">[그림6] - Pipeline #1</figcaption>
</p>
</figure>

### Pipeline #2: Independent on Repository, But Inefficient

두 번째로 구현한 Pipeline은 아래 [그림7]에서 볼 수 있습니다. Pipeline #1과 다르게 Jenkins Container가 Repository에 정의된 Dockerfile을 기반으로 Regression Test Container를 만듭니다. 이를 통해서 Repository에 의존성을 가지던 문제를 해결할 수 있었습니다. 하지만, Docker Image를 Build하는 작업은 상당히 오랜시간이 걸리기 때문에, 비효율적이라는 문제가 있었습니다.

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/7.png" alt="Gitflow Workflow" width="60%">
  <figcaption style="text-align: center;">[그림7] - Pipeline #2</figcaption>
</p>
</figure>

### Pipeline #3: InDependent on Repository, But!

첫 번째로 구현한 Pipeline은 아래 [그림8]에서 볼 수 있습니다. Docker Image는 Requirements가 변경되었을 때만 Update가 필요했습니다. 따라서 미리 DockerImage를 만들어 두고 Jenkins Container가 이를 받아서 사용하도록 구조를 변경하였습니다. Pipeline #2와 비교해봤을 때 효율적이었습니다.

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/8.png" alt="Gitflow Workflow" width="60%">
  <figcaption style="text-align: center;">[그림8] - Pipeline #3</figcaption>
</p>
</figure>

### Device Dependency

> 하지만 Pipeline #1 ~ #3은 모두 공통적으로 한 컴퓨팅 자원에 의존적이라는 문제가 있습니다. 예를 들어 Regression Test에 사용하는 컴퓨터에서 다른 작업이 돌아가고 있다면 Regression Test가 아예 작동하지 못하거나, 다른 작업을 망칠 수도 있습니다. [그림9]를 보면, 3개의 노트북이 MRX-Decktop1에 접속하여 사용하고 있는 모습을 볼 수 있습니다. 붉은 색으로 표현된 것은 남은 Memory가 많지 않다는 것을 의미합니다. 만약 Jenkins Container가 MRX-Decktop1에서 작동하고 있다면, Regression Test가 OOM(Out-of-Memory)이 발생하여 Regression Test가 정상적으로 작동할 수 없습니다.

Reviewer 2: ```suggestion
하지만 Pipeline #1 ~ #3은 모두 공통적으로 한 컴퓨팅 자원에 의존적이라는 문제가 있습니다. 예를 들어 Regression Test에 사용하는 컴퓨터에서 다른 작업이 돌아가고 있다면 Regression Test가 아예 작동하지 못하거나, 다른 작업을 망칠 수도 있습니다. [그림9]를 보면, 3개의 Process가 모두 동일한 하나의 서버에 접속하여 사용하고 있는 모습을 볼 수 있습니다. 붉은 색으로 표현된 것은 남은 Memory가 많지 않다는 것을 의미합니다. 만약 Jenkins Container도 동일한 서버에서 작동하고 있다면, Regression Test가 OOM(Out-of-Memory)이 발생하여 Regression Test가 정상적으로 작동할 수 없습니다.
```

---
layout: post
title: Regression Test, Are you sure?
author: wontak ryu
categories: [test]
image: assets/images/2020-02-10-Regression-Test/13_.gif
---

> 마키나락스는 AI Project를 넘어 AI Product로 나아가고 있습니다. Product로 나아가는 여정 속에서 재미있는 엔지니어링 이슈들이 생겼습니다.

Reviewer 3: > AI Project를 넘어 AI Product로 나아가고 있습니다.

AI Project와 AI Product 각 표현이 시사하는 바가 명확히 무엇인가요? 
(AI Product로 나아가는 것도 AI Project로 볼 수 있지 않나요?)
기술 연구개발 과정을 넘어서 제품화 과정으로 나아간다는 말씀을 하고 싶으셨던 것 걸까요?

Author: > 기술 연구개발 과정을 넘어서 제품화 과정으로 나아간다는 말씀을 하고 싶으셨던 것 걸까요?

네 맞습니다. 조금 더 명확하게 표현해보겠습니다.

> 제품에 다가갈수록 시장의 요구사항을 빠르게 반영해야하며 이런 요구사항은 계속해서 발생합니다. 이에 맞춰 Machine Learning Software도 여러가지 변화들을 빠르게 도입해야했습니다. 이를 위해서 마키나락스가 제공하는 Machine Learning Software의 성능 대해서 신속하고 정확한 검증에 대한 필요가 있었습니다. 마키나락스에서는 Unit Test는 기본적으로 적용하고 있지만, Unit Test만으로는 성능을 보장 할 수 없습니다.

Reviewer 3: 각 문장을 별도 라인으로 기술해주시면 라인단위로 리뷰를 드리기 더 편해집니다.

Author: 네 반영하겠습니다. 리뷰 감사드립니다.

> 제품에 다가갈수록 시장의 요구사항을 빠르게 반영해야하며 이런 요구사항은 계속해서 발생합니다. 이에 맞춰 Machine Learning Software도 여러가지 변화들을 빠르게 도입해야했습니다. 이를 위해서 마키나락스가 제공하는 Machine Learning Software의 성능 대해서 신속하고 정확한 검증에 대한 필요가 있었습니다. 마키나락스에서는 Unit Test는 기본적으로 적용하고 있지만, Unit Test만으로는 성능을 보장 할 수 없습니다.

Reviewer 3: Continuous Integration(지속적 통합)의 의의를 강조하고 싶으셨던게 아닐까 싶습니다. 

```suggestion
제품의 실체화에 가까워질수록 시장에서 발생하는 요구사항을 지속적으로 빠르게 반영해야 합니다. 이를 위해 마키나락스에서는 Machine Learning Software의 신속하고 정확한 성능 검증을 필요로 하게 되었습니다.
```

Reviewer 3: > 이에 맞춰 Machine Learning Software도 여러가지 변화들을 빠르게 도입해야했습니다.

앞 문장의 "빠르게 반영"이라는 표현으로 대신할 수 있을 것 같습니다.

Reviewer 3: > 마키나락스에서는 Unit Test는 기본적으로 적용하고 있지만, Unit Test만으로는 성능을 보장 할 수 없습니다.

독자분들이 "Unit Test만으로는 성능을 보장 할 수 없습니다."라는 문장을 쉽게 이해하기 어려울 것 같습니다.
Unit Test 검증만으로는 모델의 학습 성능에 대한 검증이 어려우므로... 같은 말이 덧붙여지면 더 좋겠습니다.
더 자세하면 더 좋구요.

Author: 네 반영했습니다.

이번 포스트에서는 Machine Learning Software에 대한 성능검증을 어떤 방식으로 진행하고 있는지 공유드리도록 하겠습니다.

## Problem: Can't find the cause of the lower performance!

> 우선 내부에서 AI Product를 개발하며 겪었던 문제에 대해서 공유드리도록 하겠습니다. 마키나락스에서는 GitHub를 통해 협업을 관리하고 있습니다. 따라서 개발자들은 작업이 어느 정도 완료되면 작업 중인 Branch에서 Pull-Request를 날리게 됩니다. 그럼 다른 개발자들이 해당 코드들을  Review하고 이상이 없으면 Merge 합니다.

Reviewer 3: 여기서 PR에 대한 Unit Test가 수행된다는 이야기를 언급하면 어떨까요?

Author: 반영하였습니다.

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/1.png"  width="60%">

>   <figcaption style="text-align: center;">[그림1] - Gitflow Workflow [16]</figcaption>

Reviewer 3: 레퍼런스 16번은 존재하지 않습니다.

Author: 네 반영했습니다! 감사합니다.

</p>
</figure>

> 내부에 Unit Test가 구현되어 있어서, 어느정도의 안전성이 검증된 상태였습니다. 하지만, 마키나락스는 Machine Learning Software를 다루는 조직이기 때문에 더 높은 수준의 Test System이 필요했습니다. Machine Learning의 특성상 안정성 검증뿐만 아니라 성능검증도 함께 이루어져야합니다. 현재 Test System내에서 "Unit Test는 통과하지만, 성능저하를 일으키는 Commit"같은 경우에는 추적을 할 수 없다는 문제가 있습니다. 실제로 겪었던 사례가 있습니다. Activation Function을 선택할 때, Relu를 사용하면 특정 모델의 성능저하가 일어나는 것을 발견하였습니다. Relu가 잘못된 알고리즘은 아니나, 특정모델에서는 정상적으로 작동하지 않았습니다.

Reviewer 3: ```suggestion
하지만 유닛테스트 만으로는 모델의 성능저하를 일으키는 변경사항을 알아챌 수 없다는 문제가 있었습니다. 예를 들어, 활성함수(activation function)로 ReLU를 사용했을때 특정 모델의 성능저하가 일어난 사건이 있었습니다. 이 경우 모델의 성능이 저하되었는데도 불구하고 모델의 각 모듈은 설계한대로 잘 동작했으므로 모든 테스트케이스는 문제 없이 통과했습니다.
```

  <img src="/assets/images/2020-02-10-Regression-Test/2.png"  width="60%">
  <figcaption style="text-align: center;">[그림2] - Problem in Makinarocks</figcaption>
</p>
</figure>

이런 문제가 발생했을 때, Debugging 해야할 Search Space는 [그림3]과 같습니다. 즉, 이전까지의 변화를 모두 살펴봐야 했습니다.

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/3.png"  width="60%">
  <figcaption style="text-align: center;">[그림3] - Search Space</figcaption>
</p>
</figure>

이번 문제의 경우에는 성능저하를 일으키는 원인이 두 개의 Commit이였습니다. 성능저하가 복합적인 원인들에 의해서 발생하면, 해결하는데 더 큰 어려움을 겪습니다. 두 개의 Commit을 모두 고치지 않으면, 성능저하 이슈는 계속해서 발생하기 때문입니다.

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/4.png"  width="60%">
  <figcaption style="text-align: center;">[그림4] - Causes</figcaption>
</p>
</figure>

결국, 수많은 Debugging 끝에 원인을 찾을 수 있었습니다. 그리고 원인들은 생각보다 사소한 변화였습니다. 일반적으로 생각해봤을 때 큰 문제를 야기할 것이라고 생각하기 어려운 부분이였습니다. 아마도 그렇기 때문에, 뒤늦게 발견이 되었을 것이라고 생각합니다. 더욱이 이런 Commit들이 모두 Unit Test를 통과했기 때문에, 개발자 입장에서 무엇이 원인인지 파악하기 힘듭니다.

이런 경험을 한 후에, **Are you Sure?** (이 코드 문제가 없을까요?)라는 질문에 답하기 위해서는 사실상 작은 변화더라도 **Regression Test**를 진행해야 하는 것을 깨달았습니다. 

> 여기서 Regression Test라고 정의한 것은 Machine Learning Software의 전체적인 실험 및 테스트를 진행하고 성능을 확인하는 작업을 의미합니다.

Reviewer 3: 여기서 '실험'과 '테스트'는 어떻게 다른 건가요?

Author: ```suggestion
여기서 Regression Test라고 정의한 것은 Machine Learning Software의 전체적인 학습 및 테스트를 진행하는 작업을 의미합니다.
```
표현이 모호했습니다. 이렇게 바꾸면 명확할까요?

> 변화량에 대해서 Debugging 비용을 그래프로 그려보면 [그래프1]이 나옵니다. 위에서 언급했듯이, 복합적인 원인에 의해서 성능저하 이슈가 발생했을 경우, 모든 변화의 조합에 대해서 실험 및 테스트를 진행해야합니다. (이 때, 모든 변화는 독립적이라는 가정을 하고 계산하였습니다.)

Reviewer 3: 모든 변화가 독립적이라면 변경사항 간의 의존성이 없는 것이므로 커밋의 개수만큼만 검증하면 되지 않을까요?

Author: 각각의 커밋이 서로 다른 영역의 작업이라는 표현을 하고 싶었습니다. 

모두 독립적인 경우에는 지수승만큼 검증해야하지 않을까요?
예를 들어서, a 커밋 문제, b 커밋문제가 서로 독립적이라면 a, b 둘다 해결해야 문제가 해결되지 않을까요?

만약 a + b + c 커밋이 문제를 일으킨 것이라면 이 경우에는 a or b or c 셋중에 하나만 해결되면 되는문제라서 커밋의 개수만큼 해결하면 된다고 생각했습니다.

Author: 생각해보니 불필요한 정보인거 같아 지웠습니다.

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/5.png"  width="60%">
  <figcaption style="text-align: center;">[그림5] - 경우의 수</figcaption>
</p>
</figure>
<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/graph1.png"  width="40%">
  <figcaption style="text-align: center;">[그래프1] - Cost of Debugging</figcaption>
</p>
</figure>

그럼 "언제 Regression Test를 진행해야하는가?" 라는 의문이 들 수 있습니다. [그래프1]을 근거로 내부에서는 "As Soon As Possible"(가능한 빨리)라는 결론을 내리게 되었고, 이를 위해서 쉽고 빠른 Regression Test Pipeline을 구성하였습니다.

## Trial and Errors

Regression Test Pipeline을 만들기 위해서, 여러가지 시행착오를 겪었습니다. 겪었던 시행착오를 통해서, 필요했던 **추상화 과정**에 대해서 설명드리겠습니다.

우선, 자동화 도구로 Jenkins를 활용하였습니다 [[1]](#ref-1). Jenkins는 소프트웨어 개발 시 지속적으로 통합 서비스를 제공하는 툴입니다. 비교적 높은 자유도가 있었고, 자동화 도구로 접근성이 좋다고 판단했습니다. 

### Pipeline #1: Dependent on Repository

> 첫 번째로 구현한 Pipeline은 아래 [그림6]에서 볼 수 있습니다. Jenkins Container가 Regression Test 대상이 되는 Repository의 Requirements를 미리 가지고 있습니다. 학습에 필요한 데이터의 경우 원격 저장소에 저장해두고 요청 시 접근하여 사용합니다. GitHub에서 Test요청을 보내면, Regression Test를 진행하게 됩니다. 

Reviewer 3: 독자들이 이 문장을 이해하기 위해서는 아래 용어에 대한 이해가 필요할 것 같습니다.
Jenkins Container / Repository의 Requirements

Reviewer 3: 유닛테스트, 커밋, 브랜치, 테스트 등등의 표현은 굳이 영어 표기를 사용하지 않아도 될 것 같습니다.

Author: 네 우선 Jenkins부분은 본 글에서 제외하도록 하겠습니다.

불필요한 영문이름들은 한글로 변경하겠습니다.

Author: 반영하였습니다.

  <figcaption style="text-align: center;">[그림6] - Pipeline #1</figcaption>
</p>
</figure>

### Pipeline #2: Independent on Repository, But Inefficient

두 번째로 구현한 Pipeline은 아래 [그림7]에서 볼 수 있습니다. Pipeline #1과 다르게 Jenkins Container가 Repository에 정의된 Dockerfile을 기반으로 Regression Test Container를 만듭니다. 이를 통해서 Repository에 의존성을 가지던 문제를 해결할 수 있었습니다. 하지만, Docker Image를 Build하는 작업은 상당히 오랜시간이 걸리기 때문에, 비효율적이라는 문제가 있었습니다.

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/7.png"  width="60%">
  <figcaption style="text-align: center;">[그림7] - Pipeline #2</figcaption>
</p>
</figure>

### Pipeline #3: InDependent on Repository, But!

첫 번째로 구현한 Pipeline은 아래 [그림8]에서 볼 수 있습니다. Docker Image는 Requirements가 변경되었을 때만 Update가 필요했습니다. 따라서 미리 DockerImage를 만들어 두고 Jenkins Container가 이를 받아서 사용하도록 구조를 변경하였습니다. Pipeline #2와 비교해봤을 때 효율적이었습니다.

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/8.png"  width="60%">
  <figcaption style="text-align: center;">[그림8] - Pipeline #3</figcaption>
</p>
</figure>

### Device Dependency

> 하지만 Pipeline #1 ~ #3은 모두 공통적으로 한 컴퓨팅 자원에 의존적이라는 문제가 있습니다. 예를 들어 Regression Test에 사용하는 컴퓨터에서 다른 작업이 돌아가고 있다면 Regression Test가 아예 작동하지 못하거나, 다른 작업을 망칠 수도 있습니다. [그림9]를 보면, 3개의 Process가 모두 동일한 하나의 서버에 접속하여 사용하고 있는 모습을 볼 수 있습니다. 붉은 색으로 표현된 것은 남은 Memory가 많지 않다는 것을 의미합니다. 만약 Jenkins Container도 동일한 서버에서 작동하고 있다면, Regression Test가 OOM(Out-of-Memory)이 발생하여 Regression Test가 정상적으로 작동할 수 없습니다.

Reviewer 3: ```suggestion
하지만 Pipeline #1 ~ #3은 모두 공통적으로 한 컴퓨팅 자원에 의존적이라는 문제가 있습니다. 예를 들어 Regression Test에 사용하는 컴퓨터에서 다른 작업이 돌아가고 있다면 Regression Test가 아예 작동하지 못하거나, 다른 작업을 망칠 수도 있습니다. [그림9]를 보면, 3개의 Process가 모두 동일한 하나의 서버에 접속하여 사용하고 있는 모습을 볼 수 있습니다. 붉은 색으로 표현된 것은 남은 Memory가 많지 않다는 것을 의미합니다. 만약 Jenkins Container도 동일한 서버에서 작동하고 있다면, OOM(Out-of-Memory)이 발생하여 Regression Test가 정상적으로 작동하지 않을 수 있습니다.
```

Author: 반영하였습니다.

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/9.png"  width="60%">
  <figcaption style="text-align: center;">[그림9] - Problem of Device Dependency</figcaption>
</p>
</figure>

또한 다른 MRX-Decktop2, 3를 보면 컴퓨팅 자원이 여유있다는 것을 알 수 있습니다. 자원을 효율적으로 사용하기 위해서 남은 자원에 효율적으로 접근하는 것이 필요했습니다. 이를 위해서는 Regression Test Pipeline이 특정 자원에 종속되지 않고, 필요한 자원에 동적으로 접근하여야 합니다. 위의 [그림9] 예시처럼 Regression Test Pipeline이 특정 자원의 영향을 받는 것을 개선해야합니다.

### Pipeline #4: InDependent on Device

> Device Dendency를 해결하기 위해서 Kubernetes를 사용하였습니다 [[2]](#ref-2). Kubernetes에 대해서 알고 싶으신 분들은 Reference를 참고하시는 것을 추천드립니다 [[6]](#ref-2).

Reviewer 3: ```suggestion
Device Dendency를 해결하기 위해서 Kubernetes를 사용하였습니다 [[2]](#ref-2). Kubernetes에 대해서 알고 싶으신 분들은 Kubernetes의 공식문서[[6]](#ref-2)를 참고하시는 것을 추천드립니다.
```

Author: 반영하였습니다.

Kubernetes를 사용한 목적은 내부의 컴퓨팅 자원을 추상화하기 위함입니다. 쉽게 풀어쓰면, **Kubernetes에 특정 Device를 요청하는 것이 아니라, 필요한 컴퓨팅 자원에 대해서 요청만 하면, 그에 맞는 자원할당을 받기 위해서입니다.** [그림10]을 보면, 여러가지 컴퓨팅 자원이 하나의 클러스터로 묶여있습니다. 이제 원하는 자원의 스펙을 적으면, 그에 맞는 자원이 할당될 것입니다.

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/10.png"  width="60%">
  <figcaption style="text-align: center;">[그림10] - Kubernetes in Makinarocks </figcaption>
</p>
</figure>

#### Remark: Kubernetes

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/remark_1.png" alt="Kubernetes" width="60%">
  <figcaption style="text-align: center;"> Containerization and Container Orchestration </figcaption>
</p>
</figure>

Kubernetes는 Container Orchestration을 위한 도구입니다. Container는 VM과 비교했을 때 격리 속성을 완화하여 애플리케이션 간에 운영체제를 공유합니다. 따라서 VM보다 더 가볍게 이용할 수 있습니다. Container에 대한 관심이 많아지자 자연스럽게 Container Orchestration에 대한 필요성도 강조되었습니다. Container를 쉽고 효율적으로 관리하기 위한 도구 중 하나가 Kubernetes입니다. 프로덕션 환경에서 애플리케이션을 실행하는 컨테이너가 정상작동하는지 확인하고 다운되었다면 다른 컨테이너를 실행해야합니다. Kubernetes를 활용하면 이런 문제를 시스템에 의해서 자동으로 처리할 수 있습니다. [[2]](#ref-2)
 

> 또한, Kubernetes에서는 여러 Node(Computing Resource)를 묶어서 Cluster를 구성할 수 있습니다. Cluster로 묶인 Node의 집합은 하나의 큰 Computing Resource처럼 사용될 수 있습니다. Kubernetes의 도입으로 특정 Node에 직접 접근할 필요가 없어졌고 Device Dependency 문제를 해결할 수 있었습니다.

Reviewer 3: > Device dependency 문제

쉽게 풀어서 표현하면 어떨까요?

Author: Kubernetes의 도입으로 특정 Node에 직접 접근할 필요가 없어졌고 하나의 Node에 작업이 집중되는 것을 막을 수 있었습니다.

이런식으로 변경하였습니다.

  <figcaption style="text-align: center;"> Ray Cluster </figcaption>
</p>
</figure>

Ray Cluster는 Head Node와 Worker Node로 구성됩니다. Head Node에서는 작업들을 Worker Node에 분배하여 병렬로 테스크를 수행합니다. Ray Autoscaler는 Cluster의 자원상황을 고려하여 Worker Node의 개수를 동적으로 조절할 수 있습니다. [[3]](#ref-3)

Jenkins Container의 역할은 특정 Device내에서 Container로 Regression Test를 진행하는 것이 아닙니다. Jenkins Container는 미리 정의된 컴퓨팅 자원 스펙에 해당하는 Ray Cluster를 만드는 것입니다 [[3]](#ref-2). 여기서 Ray Cluster의 역할은 Regression Test를 병렬적으로 진행하기 위한 목적으로 사용되고, 작업이 끝나게 되면 Ray Cluster는 사라지게 됩니다. 참고로 [그림10]에서 구성한 Cluster와 Ray Cluster는 다른 역할을 합니다. [그림10]은 자원자체를 묶는 작업을 의미한다면, Ray Cluster는 이미 묶인 자원을 활용하는 것입니다. 

이제 Kubernetes 그리고 Ray Cluster를 활용하여, [그림11]과 같은 Pipeline을 구축하였습니다. Repository에 의존성을 제거하였으며, Docker Image도 미리 만들어둔 Image를 활용하였습니다. 또한, Device에 대한 의존성을 제거하여, 내부의 컴퓨팅 자원을 더욱 효율적으로 사용할 수 있습니다.

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/11.png"  width="60%">
  <figcaption style="text-align: center;">[그림11] - Pipeline #4 </figcaption>
</p>
</figure>

## (Selected) Method: Self-Hosted Runner in GitHub Action

> Pipeline #1부터 #4까지 모두 Jenkins를 사용하고 있습니다. 하지만 Jenkins라는 툴에 익숙하지 않다보니 기술적인 이슈가 발생했을 때 대처하는데 쉽지 않았습니다. 특히 Jenkins에서 Kubernetes의 Application을 활용하려면 별도의 Plugin설치와 문법을 익혀야합니다. 조금 더 쉬운 방법이 없을까 고민이 들었습니다.

Reviewer 3: 앞 내용을 힘겹게 따라온 독자들이 다소 허무할 수도 있을 것 같네요.
젠킨스 내용을 줄이고 깃헙 액션에 대한 비중을 높이면 어떨까요?

Author: 네 리뷰 포인트 감사드립니다!

                stage('Build a Maven project') {
                    sh 'mvn -B clean install'
                }
            }
        }

        stage('Get a Golang project') {
            git url: 'https://github.com/hashicorp/terraform.git'
            container('golang') {
                stage('Build a Go project') {
                    sh """
                    mkdir -p /go/src/github.com/hashicorp
                    ln -s `pwd` /go/src/github.com/hashicorp/terraform
                    cd /go/src/github.com/hashicorp/terraform && make core-dev
                    """
                }
            }
        }

    }
}
```

그러던 중 GitHub Action에서 Self-Hosted Runner라는 서비스를 제공하는 것을 발견했습니다 [[4]](#ref-2). Self-Hosted-Runner는 가지고 있는 자원을 통해서 Github Action 진행할 수 있었습니다. 상대적으로 GitHub에서 관련내용에 대해서 문서를 제공하였고, 문법도 직관적이라고 생각이 들었습니다. 이런 특징들은 유지보수 관점에서 높은 점수를 줄 수 있었고, 기존의 Jenkins의 역할을 GitHub Action으로 대체하기로 하였습니다.

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/12.png"  width="60%">
  <figcaption style="text-align: center;">[그림12] - Pipeline #5 </figcaption>
</p>
</figure>

## Are You Sure? Yes!

> GitHub Action에서 Trigger Event Type에 대해서 정할 수 있습니다. 여러 논의 끝에, Workflow Dispatch라는 Type을 선택하였습니다. 이제 마우스 클릭으로 GitHub Web에서 Regression Test를 실행할 수 있습니다 [[5]](#ref-2).

Reviewer 3: 이 문장을 이해하기 위한 진입장벽이 다소 높아보입니다.

1. GitHub Action을 알아야함
2. GitHub Action의 Trigger Event Type 지정 방법을 알아야 함
3. Workflow Dispatch가 무엇인지 알아야 함

Author: jenkins 비중을 줄이고 Action에 대한 비중을 높여보겠습니다.

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/13.gif"  width="60%">
  <figcaption style="text-align: center;">[그림13] - Click for Regression Test </figcaption>
</p>
</figure>

> Regression Test Pipeline의 모습을 [그림13]으로 도식화해봤습니다. GitHub에서 미리 설정한 Event Type에 해당하는 Event가 발생하면, MRX-Hosted-Runner에게 Regression Test를 요청합니다. MRX-Hosted-Runner는 Ray Cluster를 구성합니다. 학습 및 실험을 진행할 때는 MLflow에 실험정보를 로깅하고, 학습이 끝나면 이에 대한 정보를 GitHub에 전달합니다. 현재는 해당 PR에 Comment로 MLflow 실험링크를 달아주는 방식으로 사용중입니다.

Reviewer 3: MLflow를 모르는 사람은 이해하기 어려울 것 같습니다.

Author: MLflow에 부가적으로 중앙화된 실험결과 로깅 서비스라는 것을 명시해보겟습니다.

Author: Reference를 추가하였습니다.

Regression Test Pipeline을 구성하기 위해서, 여러가지 추상화과정이 필요했습니다. Repository에 독립적으로 작동할 수 있어야 했습니다. 또한, Machine Learning Software(AI)는 많은 컴퓨팅 자원을 요구하기 때문에, 효율적인 자원사용이 필요했습니다. 이를 위해서 Kubernetes를 활용하여 컴퓨팅 자원을 가상화하였습니다. 

Regression Test를 통해서 Search Space를 줄일 수 있었고, $2^\text{ReducedSearchSpace}$ 만큼의 Debugging Cost를 줄일 수 있었습니다. 그리고 클릭 한 번으로 실험을 진행할 수 있다는 것도 굉장히 매력적인 일이였습니다.

이번 포스트를 통해서 비슷한 문제를 고민하는 분들께 작은 도움이 되었으면 좋겠습니다.

## Reference

<a name="ref-1">[1]</a>  [jenkins[websites], (2020, Feb, 10)](https://www.jenkins.io/)

<a name="ref-2">[2]</a>  [Kubernetes[websites], (2020, Feb, 10)](https://kubernetes.io/)

<a name="ref-3">[3]</a>  [Ray cluster[websites], (2020, Feb, 10)](https://docs.ray.io/en/master/cluster/index.html)

<a name="ref-4">[4]</a>  [About Self Hosted Runners[websites], (2020, Feb, 10)](https://docs.github.com/en/actions/hosting-your-own-runners/about-self-hosted-runners)

<a name="ref-5">[5]</a>  [Workflow Dispatch[websites], (2020, Feb, 10)](https://docs.github.com/en/actions/reference/events-that-trigger-workflows#workflow_dispatch)

> <a name="ref-6">[6]</a>  [What is Kubernetes[websites], (2020, Feb, 10)](https://www.google.com/search?q=what+is+kubernetes&oq=what+is+kubernetes&aqs=chrome..69i57j0l5j69i60l2.7878j0j4&sourceid=chrome&ie=UTF-8)

Reviewer 3: 링크가 잘못 들어간 것 같습니다.
의도하신 것은 아래 링크 일까요?

https://kubernetes.io/ko/docs/concepts/overview/what-is-kubernetes/

Author: 감사합니다. 반영했습니다.

> <a name="ref-6">[6]</a>  [What is Kubernetes[websites], (2020, Feb, 10)](https://www.google.com/search?q=what+is+kubernetes&oq=what+is+kubernetes&aqs=chrome..69i57j0l5j69i60l2.7878j0j4&sourceid=chrome&ie=UTF-8)

Reviewer 3: Which referencing style is this?

Author: http://edu.copykiller.com/edu-source/faq/?mod=document&uid=156

Author: ![image](https://user-images.githubusercontent.com/27891090/111436641-8f346180-8745-11eb-8339-6a71b53792fd.png)

Author: 여기서 보고 했습니다.

Reviewer 3: Check this out too
https://libguides.murdoch.edu.au/APA/internet

---
layout: post
title: Regression Test, Are you sure?
author: wontak ryu
categories: [test]
image: assets/images/2020-02-10-Regression-Test/13_.gif
---

마키나락스는 AI 기술개발을 넘어서 AI 제품화과정으로 나아가고 있습니다. 
제품화과정으로 나아가는 여정 속에서 재미있는 엔지니어링 이슈들이 생겼습니다.
제품의 실체화에 가까워질수록 시장에서 발생하는 요구사항을 지속적으로 빠르게 반영해야 합니다.
이를 위해 Machine Learning Software의 신속하고 정확한 성능 검증을 필요로 하게 되었습니다.
유닛테스트 검증만으로 모델의 학습성능에 대한 검증이 어려우므로, 성능검증에 대한 다른 방법이 필요했습니다.

이번 포스트에서는 Machine Learning Software에 대한 성능검증을 어떤 방식으로 진행하고 있는지 공유드리겠습니다.

## Problem: Can't find the cause of the lower performance!

우선 AI Product를 개발하며 겪었던 문제에 대해서 공유드리겠습니다. 
GitHub를 통해 협업을 관리하고 있습니다. 

> 개발자들은 작업이 어느 정도 완료되면 작업 중인 브랜치에서 Pull-Request를 날리게 됩니다. 
> 유닛테스트로 코드가 제대로 작동하는지 검증합니다. 

Reviewer 3: `날리게 됩니다` sounds bit of informal. People use `open` as a verb for PR in formal English.
e.g.
>  Once a pull request is opened, you can discuss and review the potential changes with collaborators and add follow-up commits before your changes are merged into the base branch.

```suggestion
개발자들은 작업이 어느 정도 완료되면 Pull Request를 통해 작업 내용을 푸쉬하게 되고, 이 과정에서 유닛테스트를 통해 코드의 검증을 시행합니다.
```

Author: 감사합니다.

> 개발자들은 작업이 어느 정도 완료되면 작업 중인 브랜치에서 Pull-Request를 날리게 됩니다. 
> 유닛테스트로 코드가 제대로 작동하는지 검증합니다. 
> 검증 후에 다른 개발자들이 해당 코드들을  Review하고 이상이 없으면 Merge 합니다.

Reviewer 3: or you can list the procedure.

```suggestion
개발자들은 작업이 어느 정도 완료되면

1.  Pull-Request를 통해 작업내용을 푸쉬합니다.
2. 유닛테스트로 코드를 검증합니다.
3. 작업내용을 동료들이 리뷰합니다.
4. 코드를 병합(Merge)합니다.
```

Author: 이런 식의 표현이 더 명료해보입니다. 반영하겠습니다.

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/1.png"  width="60%">
  <figcaption style="text-align: center;">[그림1] - Gitflow Workflow</figcaption>
</p>

> </figure>

Reviewer 3: Add reference
https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow

Author: 반영했습니다.

하지만 유닛테스트 만으로는 모델의 성능저하를 일으키는 변경사항을 알아챌 수 없다는 문제가 있었습니다. 
예를 들어, 활성함수(activation function)로 ReLU를 사용했을때 특정 모델의 성능저하가 일어난 사건이 있었습니다. 
이 경우 모델의 성능이 저하되었는데도 불구하고 각 모듈은 설계한대로 잘 동작했으므로 모든 테스트케이스는 문제 없이 통과했습니다.

아래의 [그림2]는 실제로 겪었던 문제입니다. 여러 브랜치가 Merge된 상태에서 origin/master의 성능저하를 발견하였습니다. 

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/2.png"  width="60%">

>   <figcaption style="text-align: center;">[그림2] - Problem in Makinarocks</figcaption>

Reviewer 3: [그림2] 자체가 problem인 것은 아니므로 `문제상황 당시의 작업흐름` 같은 표현을 사용하는 것은 어떨까요?

Reviewer 3: [그림2]의 핵심인 Low Performance가 한 눈에 잘 들어오지 않습니다.
이 부분의 폰트를 highlighting 하면 어떨까요?

Author: 반영하였습니다. 감사합니다.

</p>
</figure>

> 이런 문제가 발생했을 때, 디버깅 해야할 Search Space는 [그림3]과 같습니다.
> 즉, 이전까지의 변화를 모두 살펴봐야 했습니다.

Reviewer 3: ```suggestion
이때 어느 커밋(commit)이 성능 저하를 유발했는지 알 수 없었으므로 성능저하를 발견한 시점 이전의 변경사항을 모두 살펴봐야 했습니다.
즉, 당시 디버깅 해야할 Search Space는 [그림3]과 같았습니다.
```

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/3.png"  width="60%">
  <figcaption style="text-align: center;">[그림3] - Search Space</figcaption>
</p>
</figure>

이번 문제의 경우에는 성능저하를 일으키는 원인이 두 개의 커밋이였습니다. 
성능저하가 복합적인 원인들에 의해서 발생하면 해결하는데 더 큰 어려움을 겪습니다. 

> 두 개의 커밋을 모두 고치지 않으면 성능저하 이슈는 계속해서 발생하기 때문입니다.

Reviewer 3: 흐름상 어려움을 겪는 가장 큰 요인은 넓은 탐색공간에서 성능저하의 원인을 파악해야 하기 때문 아닐까요?

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/4.png"  width="60%">
  <figcaption style="text-align: center;">[그림4] - Causes</figcaption>
</p>

> </figure>

Reviewer 3: Cause의 폰트를 highlighting 하면 어떨까요?

Author: 반영하였습니다.

  <figcaption style="text-align: center;">[그림5] - 경우의 수</figcaption>
</p>
</figure>
<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/graph1.png"  width="40%">
  <figcaption style="text-align: center;">[그래프1] - 디버깅 비용</figcaption>
</p>
</figure>

그럼 "언제 Regression Test를 진행해야하는가?" 라는 의문이 들 수 있습니다.
[그래프1]을 근거로 "As Soon As Possible"(가능한 빨리)라는 결론을 내리게 되었고 이를 위해서 쉽고 빠른 Regression Test Pipeline을 구성하였습니다.

## Trial and Errors

Regression Test Pipeline을 만들기 위해서 여러가지 시행착오를 겪었습니다. 
겪었던 시행착오를 통해서 필요했던 **추상화 과정**에 대해서 설명드리겠습니다.

우선 자동화 도구로 GitHub Actions의 Self-Hosted Runner를 활용하였습니다 [[1]](#ref-1). 
Self-Hosted Runner는 내부 자원으로 GitHub Actions의 기능들을 사용할 수 있도록 지원하는 도구입니다.

#### Remark: GitHub Actions and Self-Hosted Runner

GitHub Actions는 자동화된 CI/CD 기능들을 제공하며 이를 커스터마이즈하여 활용할 수 있는 도구입니다.
더 자세한 정보를 알고 싶으신 분들은 GitHub Actions의 Quickstart for GitHub Actions[[8]](#ref-8)을 참고하시기 바랍니다.

GitHub Actions은 사용자가 `./github/workflows` 디렉토리에 GitHub Action를 위한 yml파일을 넣게 되면 작동합니다.
yml파일에 어떤 Runner를 사용할지 결정할 수 있는데, Ubuntu, Mac OS, Windows Server등 다양한 환경을 선택할 수 있습니다.
Runner가 선택되면 GitHub에서는 가상환경을 만들어 정해진 테스크를 수행합니다.

> 이 때 사용되는 자원은 GitHub에서 제공합니다.
> 사용량에 따른 과금정책을 가지고 있습니다.

Reviewer 3: ```suggestion
이 때의 컴퓨팅 자원은 GitHub에서 제공되는데 자원의 사용량에 따른 [과금정책](https://github.com/pricing)을 가지고 있습니다.
```

Author: 반영했습니다.

> Self-Hosted Runner는 내부 자원을 사용하여 가상환경을 만듭니다. [[4]](#ref-4)

Reviewer 3: ```suggestion
Self-Hosted Runner는 내부 자원을 사용하여 가상환경을 만듭니다 [[4]](#ref-4).
```

컴퓨팅 리소스가 많이 사용될 때 유용합니다. 

> 특히 Regression Test는 GPU자원을 사용해야하기 때문에 매우 유용했습니다.

Reviewer 3: 전반적으로 문장 간의 연결이 자연스럽지 않은 부분들이 많아 보입니다.

Author: 반영했습니다.

> 이렇게 만들어진 Self-Hosted Runner는 아래와 같이 선택할 수 있습니다.

Reviewer 3: '이렇게 만들어진'이라고 표현하기에는 어떻게 만드는지에 대한 내용이 부족해보입니다.

Author: 반영했습니다.

```yml
name: Regression_Test

jobs:
  regression_test:
    runs-on: [self-hosted, ubuntu-18.04]
```

### Pipeline #1: Dependent on Repository

첫 번째로 구현한 Pipeline은 아래 [그림6]에서 볼 수 있습니다. 
MRX-Hosted Runner가 Regression Test 대상이 되는 Repository의 Requirements(필요환경)를 미리 가지고 있습니다. 
학습에 필요한 데이터의 경우 원격 저장소에 저장해두고 요청 시 접근하여 사용합니다. 
GitHub에서 테스트요청을 보내면 Regression Test를 진행하게 됩니다. 

이런 구조는 한 Repository에 의존성을 가지게 된다는 문제를 가지고 있습니다.

> 특정 Repository를 위한 MRX-Hosted Runner가 다른 Repository를 운영할 수 없습니다.

Reviewer 3: 이해를 돕기 위해 예시가 있었으면 좋겠습니다.

</p>
</figure>

### Pipeline #2: Independent on Repository, But Inefficient

두 번째로 구현한 Pipeline은 아래 [그림7]에서 볼 수 있습니다. 
Pipeline #1과 다르게 MRX-Hosted Runner가 Repository에 정의된 Dockerfile을 기반으로 Regression Test Container를 만듭니다. 
이를 통해서 Repository에 의존성을 가지던 문제를 해결할 수 있었습니다. 
하지만 Docker Image를 Build하는 작업은 상당히 오랜시간이 걸리기 때문에 비효율적이라는 문제가 있었습니다.

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/7.png"  width="60%">
  <figcaption style="text-align: center;">[그림7] - Pipeline #2</figcaption>
</p>
</figure>

### Pipeline #3: InDependent on Repository, But!

첫 번째로 구현한 Pipeline은 아래 [그림8]에서 볼 수 있습니다. 
Docker Image는 Requirements가 변경되었을 때만 Update가 필요했습니다. 
따라서 미리 DockerImage를 만들어 두고 MRX-Hosted Runner가 이를 받아서 사용하도록 구조를 변경하였습니다. 
Pipeline #2와 비교해봤을 때 효율적이었습니다.

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/8.png"  width="60%">
  <figcaption style="text-align: center;">[그림8] - Pipeline #3</figcaption>
</p>
</figure>

### Device Dependency

하지만 Pipeline #1 ~ #3은 모두 공통적으로 한 컴퓨팅 자원에 의존적이라는 문제가 있습니다.

> 예를 들어 Regression Test에 사용하는 컴퓨터에서 다른 작업이 돌아가고 있다면 Regression Test가 아예 작동하지 못하거나 다른 작업을 망칠 수도 있습니다.

Reviewer 3: ```suggestion
예를 들어 Regression Test에 사용하는 컴퓨터에서 어떤 작업을 수행하고 있다면 Regression Test의 요청이 수락되지 않거나 수행중이던 작업에 영향을 줄 수도 있습니다.
```

Author: 반영했습니다. 감사합니다.

[그림9]를 보면 3개의 Process가 모두 동일한 하나의 서버에 접속하여 사용하고 있는 모습을 볼 수 있습니다.
붉은 색으로 표현된 것은 남은 Memory가 많지 않다는 것을 의미합니다.
만약 MRX-Hosted Runner도 동일한 서버에서 작동하고 있다면 OOM(Out-of-Memory)가 발생하여 Regression Test가 정상적으로 작동하지 않을 수 있습니다.

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/9.png"  width="60%">
  <figcaption style="text-align: center;">[그림9] - Problem of Device Dependency</figcaption>
</p>
</figure>

또한 다른 MRX-Decktop2, 3를 보면 컴퓨팅 자원이 여유있다는 것을 알 수 있습니다. 
자원을 효율적으로 사용하기 위해서 남은 자원에 효율적으로 접근하는 것이 필요했습니다. 
이를 위해서는 Regression Test Pipeline이 특정 자원에 종속되지 않고 필요한 자원에 동적으로 접근하여야 합니다. 

> 위의 [그림9] 예시처럼 Regression Test Pipeline이 특정 자원의 영향을 받는 것을 개선해야합니다.

Reviewer 3: ```suggestion
그러기 위해서는 위의 [그림9] 예시처럼 Regression Test Pipeline이 특정 자원의 영향을 받는 것을 개선해야합니다.
```

Author: ```suggestion
즉, 위의 [그림9] 예시처럼 Regression Test Pipeline이 특정 자원의 영향을 받는 것을 개선해야합니다.
```
유사한 의미의 문구라서 위와 같이 수정하였습니다.

### Pipeline #4: InDependent on Device

Device Dendency를 해결하기 위해서 Kubernetes를 사용하였습니다 [[2]](#ref-2).
Kubernetes에 대해서 알고 싶으신 분들은 Kubernetes의 공식문서[[6]](#ref-2)를 참고하시는 것을 추천드립니다.

Kubernetes를 사용한 목적은 내부의 컴퓨팅 자원을 추상화하기 위함입니다. 
쉽게 풀어쓰면, **Kubernetes에 특정 Device를 요청하는 것이 아니라, 필요한 컴퓨팅 자원에 대해서 요청만 하면, 그에 맞는 자원할당을 받기 위해서입니다.** 
[그림10]을 보면 여러가지 컴퓨팅 자원이 하나의 클러스터로 묶여있습니다. 
이제 원하는 자원의 스펙을 적으면, 그에 맞는 자원이 할당될 것입니다.

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/10.png"  width="60%">
  <figcaption style="text-align: center;">[그림10] - Kubernetes in Makinarocks </figcaption>
</p>
</figure>

#### Remark: Kubernetes

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/remark_1.png" alt="Kubernetes" width="60%">
  <figcaption style="text-align: center;"> Containerization and Container Orchestration </figcaption>
</p>

> </figure>

Reviewer 3: 그림 번호 및 출처없음

Author: 반영하였습니다.

Kubernetes는 Container Orchestration을 위한 도구입니다. 
Container는 VM과 비교했을 때 격리 속성을 완화하여 애플리케이션 간에 운영체제를 공유합니다. 
따라서 VM보다 더 가볍게 이용할 수 있습니다. 
Container에 대한 관심이 많아지자 자연스럽게 Container Orchestration에 대한 필요성도 강조되었습니다. 
Container를 쉽고 효율적으로 관리하기 위한 도구 중 하나가 Kubernetes입니다. 
프로덕션 환경에서 애플리케이션을 실행하는 컨테이너가 정상작동하는지 확인하고 다운되었다면 다른 컨테이너를 실행해야합니다. 
Kubernetes를 활용하면 이런 문제를 시스템에 의해서 자동으로 처리할 수 있습니다. [[2]](#ref-2)
 
또한 Kubernetes에서는 여러 노드(컴퓨팅 자원)를 묶어서 Cluster를 구성할 수 있습니다.
Cluster로 묶인 노드의 집합은 하나의 큰 컴퓨팅 자원처럼 사용될 수 있습니다. 
Kubernetes의 도입으로 특정 노드에 직접 접근할 필요가 없어졌고 하나의 노드에 작업이 집중되는 것을 막을 수 있었습니다.

#### Remark: Ray Cluster

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/remark_2.png" alt="Kubernetes" width="60%">
  <figcaption style="text-align: center;"> Ray Cluster </figcaption>
</p>

> </figure>

Reviewer 3: 그림 번호 및 출처없음

Author: 반영하였습니다.

Regression Test를 하는데 너무 많은 시간이 소요된다면 업무의 병목이 될 수 있습니다.
Ray Cluster는 병렬로 실험을 진행하기 때문에 Regression Test에 걸리는 시간을 줄일 수 있습니다.

Ray Cluster는 헤드 노드와 워커 노드로 구성됩니다. 
헤드 노드에서는 작업들을 워커 노드에 분배하여 병렬로 테스크를 수행합니다. 
Ray Cluster는 작업들을 병렬적으로 처리하여 Regression Test를 빠르게 진행할 수 있습니다.

> Ray Autoscaler는 Cluster의 자원상황을 고려하여 워커 노드의 개수를 동적으로 조절할 수 있습니다. [[3]](#ref-3)

Reviewer 3: ```suggestion
Ray Autoscaler는 Cluster의 자원상황을 고려하여 워커 노드의 개수를 동적으로 조절할 수 있습니다 [[3]](#ref-3).
```

Author: 감사합니다.

MRX-Hosted Runner의 역할은 특정 Device내에서 Container로 Regression Test를 진행하는 것이 아닙니다. 
미리 정의된 컴퓨팅 자원 스펙에 해당하는 Ray Cluster를 만드는 것입니다 [[3]](#ref-2). 
여기서 Ray Cluster의 역할은 Regression Test를 병렬적으로 진행하기 위한 목적으로 사용되고 작업이 끝나게 되면 Ray Cluster는 사라지게 됩니다. 
참고로 [그림10]에서 구성한 Cluster와 Ray Cluster는 다른 역할을 합니다. 
[그림10]은 자원자체를 묶는 작업을 의미한다면 Ray Cluster는 이미 묶인 자원을 활용하는 것입니다. 

이제 Kubernetes 그리고 Ray Cluster를 활용하여 [그림11]과 같은 Pipeline을 구축하였습니다. 
Repository에 의존성을 제거하였으며 Docker Image도 미리 만들어둔 Image를 활용하였습니다. 

> 또한 Device에 대한 의존성을 제거하여 내부의 컴퓨팅 자원을 더욱 효율적으로 사용할 수 있습니다.

Reviewer 3: 연결된 앞문장과의 시제일치
```suggestion
또한 Device에 대한 의존성을 제거하여 내부의 컴퓨팅 자원을 더욱 효율적으로 사용할 수 있었습니다.
```

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/11.png"  width="60%">
  <figcaption style="text-align: center;">[그림11] - Pipeline #4 </figcaption>
</p>
</figure>

## Are You Sure? Yes!

GitHub Action에서 Trigger Event Type에 대해서 정할 수 있습니다. 
여러 논의 끝에, Workflow Dispatch라는 Type을 선택하였습니다.

Workflow Dispatch를 활용하면 Regression Test를 원하는 순간에 쉽게 진행할 수 있었습니다.

#### Remark: GitHub Actions Event Type

GitHub Actions는 특정 이벤트에 대하여 정해진 테스크를 수행할 수 있습니다.
예를 들어 Pull Request 이벤트가 발생할 때 테스크를 수행하고 싶다면 아래와 같이 작성하면 됩니다.

```yml
name: Regression_Test
on:
  pull_request:
```
Workflow Dispatch는 선택적으로 GitHub Action을 수행하고 싶을 때 사용합니다 [[5]](#ref-5).
매 Pull Request 혹은 Push마다 Regression Test를 수행한다면 너무 많은 실험을 진행해야합니다.
코드리뷰가 끝난 후에 Regression Test를 수행하기 위하여 Workflow Dispatch를 선택하였습니다.
Workflow Dispatch를 사용하기 위해서는 아래와 같이 작성하면 됩니다.

```yml
name: Regression_Test
on:

>   workflow_dispatch:

Reviewer 3: workflow_dispatch를 어떻게 정의했는지는 따로 다루지 않을 예정인가요?

Author: 네 workflow dispatch의 디테일한 내용은 다루지 않을려고 합니다.

현재는 실행여부만 결정합니다.

```

> 이제 마우스 클릭으로 GitHub Web에서 Regression Test를 실행할 수 있습니다.

Reviewer 3: 코드리뷰가 완료된 뒤에 자동으로 Regression Test가 수행되는 것이 아니라 수작업으로 테스트를 시행하는 건가요?

Author: 네 manually 하게 작업하고 있습니다.

Author: 자동화 부분도 추가하였습니다.

유닛테스트 검증만으로 모델의 학습성능에 대한 검증이 어려우므로, 성능검증에 대한 다른 방법이 필요했습니다.

이번 포스트에서는 Machine Learning Software에 대한 성능검증을 어떤 방식으로 진행하고 있는지 공유드리겠습니다.

## Problem: Can't find the cause of the lower performance!

우선 AI Product를 개발하며 겪었던 문제에 대해서 공유드리겠습니다. 
GitHub를 통해 협업을 관리하고 있습니다.

개발자들은 작업을 완료하면 다음의 과정을 통해 코드를 병합(Merge)합니다.

1. Pull-Request를 통해 작업내용을 푸쉬합니다.
2. 유닛테스트로 코드를 검증합니다.
3. 작업내용을 동료들이 리뷰합니다.
4. 코드를 병합니다.

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/1.png"  width="60%">
  <figcaption style="text-align: center;">[그림1] - Gitflow Workflow [1]</figcaption>
</p>
</figure>

하지만 유닛테스트 만으로는 모델의 성능저하를 일으키는 변경사항을 알아챌 수 없다는 문제가 있었습니다. 
예를 들어, 활성함수(activation function)로 ReLU를 사용했을때 특정 모델의 성능저하가 일어난 사건이 있었습니다. 
이 경우 모델의 성능이 저하되었는데도 불구하고 각 모듈은 설계한대로 잘 동작했으므로 모든 테스트케이스는 문제 없이 통과했습니다.

아래의 [그림2]는 실제로 겪었던 문제입니다. 여러 브랜치가 Merge된 상태에서 origin/master의 성능저하를 발견하였습니다. 

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/2.png"  width="60%">
  <figcaption style="text-align: center;">[그림2] - 문제상황 당시의 작업흐름</figcaption>
</p>
</figure>

> 이때 어느 병합(Merge)가 성능 저하를 유발했는지 알 수 없었으므로 성능저하를 발견한 시점 이전의 변경사항을 모두 살펴봐야 했습니다.

Reviewer 3: ```suggestion
이때 어느 병합(Merge)이 성능 저하를 유발했는지 알 수 없었으므로 성능저하를 발견한 시점 이전의 변경사항을 모두 살펴봐야 했습니다.
```

즉, 당시 디버깅 해야할 커밋은 [그림3]에서 붉은색으로 음영처리된 부분이였습니다.

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/3.png"  width="60%">
  <figcaption style="text-align: center;">[그림3] - 디버깅 대상의 커밋</figcaption>
</p>
</figure>

> 이전 히스토리를 모두 살펴보는 작업은 매우 소모적입니다.

Reviewer 3: ```suggestion
이전 히스토리를 차례대로 검증하는 작업은 매우 소모적입니다.
```

> 게다가 기민하게 시장의 요구사항을 반영해야하는 상황에서 디버깅 작업은 팀에 부담을 줍니다.

Reviewer 3: ```suggestion
게다가 기민하게 시장의 요구사항을 반영해야하는 상황에서 늘어지는 디버깅 작업은 팀에 부담을 줍니다.
```

결국 수많은 디버깅 끝에 원인을 찾을 수 있었습니다. 
그리고 원인들은 생각보다 사소한 변화였습니다. 
일반적으로 생각했을때 큰 문제를 야기할 것이라고 생각하기 어려운 부분이였습니다. 
아마도 그렇기 때문에 뒤늦게 발견이 되었을 것이라고 생각합니다. 

> 더욱이 이런 커밋들이 모두 유닛테스트를 통과했기 때문에 개발자 입장에서 무엇이 원인인지 파악하기 힘듭니다.

Reviewer 3: ```suggestion
더욱이 이런 커밋들이 모두 유닛테스트를 통과했기 때문에 개발자 입장에서 무엇이 원인인지 파악하기 힘들었습니다.
```
시제일치

이런 경험을 한 후에 **Are you Sure?** (이 코드 문제가 없을까요?)라는 질문에 답하기 위해서는 사실상 작은 변화더라도 **Regression Test**를 진행해야 하는 것을 깨달았습니다. 

여기서 Regression Test라고 정의한 것은 Machine Learning Software의 전체적인 학습 및 테스트를 진행하고 성능을 확인하는 작업을 의미합니다.

문제는 빨리 발견할수록 해결하기 쉽습니다.

> 문제를 빨리 발견하기 위해서 Regression Test를 "As Soon As Possible"(가능한 빨리)진행해야 합니다.

Reviewer 3: ```suggestion
문제를 빨리 발견하기 위해서 Regression Test를 "As Soon As Possible"(가능한 빨리) 진행해야 합니다.
```

> 이를 위해서 쉽고 빠른 Regression Test Pipeline을 구성하였습니다.

Reviewer 3: `쉽고 빠른`의 대상이 무엇인가요?
1. Regression Test Pipeline
2. Regression Test Pipeline 구측

Author: ```suggestion
이를 위해서 쉽고 빠르게 Regression Test를 할 수 있도록 Regression Test Pipeline을 구성하였습니다.
```
이런식으로 수정했습니다.

## Trial and Errors

Regression Test Pipeline을 만들기 위해서 여러가지 시행착오를 겪었습니다. 

> 겪었던 시행착오를 통해서 필요했던 **추상화 과정**에 대해서 설명드리겠습니다.

Reviewer 3: 문맥 속에서 추상화 과정이 의미하는 바가 무엇인지 잘 모르겠습니다.
좀 더 풀어서 써주시면 좋겠습니다.

Author: ```suggestion
앞으로 소개할 과정들은 Pipeline 구축시 다른 요소에 의존성이 있던 부분을 제거하거나 효율화하는 내용입니다.
```
네 리뷰 감사드립니다.

우선 자동화 도구로 GitHub Actions의 Self-Hosted Runner를 활용하였습니다 [[4]](#ref-4). 
Self-Hosted Runner는 내부 자원으로 GitHub Actions의 기능들을 사용할 수 있도록 지원하는 도구입니다.

#### Remark: GitHub Actions and Self-Hosted Runner

> GitHub Actions는 자동화된 CI/CD 기능들을 제공하며 이를 커스터마이즈하여 활용할 수 있는 도구입니다.

Reviewer 3: 1. CI/CD 안에 자동화의 의미가 내포되어 있습니다.
2. CI/CD는 GitHub Actions를 이용하여 수행할 수 있는 일부 항목입니다. 용어 설명은 가급적 공식적인 자료를 참고하면 좋겠습니다.

```
Automate, customize, and execute your software development workflows right in your repository with GitHub Actions. You can discover, create, and share actions to perform any job you'd like, including CI/CD, and combine actions in a completely customized workflow.
```

https://docs.github.com/en/actions

Author: 네 감사합니다.

더 자세한 정보를 알고 싶으신 분들은 GitHub Actions의 Quickstart for GitHub Actions[[8]](#ref-8)을 참고하시기 바랍니다.

> GitHub Actions은 사용자가 `./github/workflows` 디렉토리에 GitHub Action를 위한 yml파일을 넣게 되면 작동합니다.

Reviewer 3: ```suggestion
GitHub Actions는 사용자가 `./github/workflows` 디렉토리에 GitHub Action를 위한 yml파일을 넣게 되면 작동합니다.
```

Kubernetes에 대해서 알고 싶으신 분들은 Kubernetes의 공식문서[[6]](#ref-2)를 참고하시는 것을 추천드립니다.

Kubernetes를 사용한 목적은 내부의 컴퓨팅 자원을 추상화하기 위함입니다. 
쉽게 풀어쓰면, **Kubernetes에 특정 Device를 요청하는 것이 아니라, 필요한 컴퓨팅 자원에 대해서 요청만 하면, 그에 맞는 자원할당을 받기 위해서입니다.** 
[그림8]을 보면 여러가지 컴퓨팅 자원이 하나의 클러스터로 묶여있습니다. 
이제 원하는 자원의 스펙을 적으면, 그에 맞는 자원이 할당될 것입니다.

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/8.png"  width="60%">
  <figcaption style="text-align: center;">[그림8] - Kubernetes in Makinarocks </figcaption>
</p>
</figure>

#### Remark: Kubernetes

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/9.png" alt="Kubernetes" width="60%">
  <figcaption style="text-align: center;"> [그림9] - Containerization and Container Orchestration [6] </figcaption>
</p>
</figure>

Kubernetes는 Container Orchestration을 위한 도구입니다. 
Container는 VM과 비교했을 때 격리 속성을 완화하여 애플리케이션 간에 운영체제를 공유합니다. 
따라서 VM보다 더 가볍게 이용할 수 있습니다. 
Container에 대한 관심이 많아지자 자연스럽게 Container Orchestration에 대한 필요성도 강조되었습니다. 
Container를 쉽고 효율적으로 관리하기 위한 도구 중 하나가 Kubernetes입니다. 
프로덕션 환경에서 애플리케이션을 실행하는 컨테이너가 정상작동하는지 확인하고 다운되었다면 다른 컨테이너를 실행해야합니다. 

> Kubernetes를 활용하면 이런 문제를 시스템에 의해서 자동으로 처리할 수 있습니다. [[2]](#ref-2)

Reviewer 3: ```suggestion
Kubernetes를 활용하면 이런 문제를 시스템에 의해서 자동으로 처리할 수 있습니다 [[2]](#ref-2).
```

 
또한 Kubernetes에서는 여러 노드(컴퓨팅 자원)를 묶어서 Cluster를 구성할 수 있습니다.
Cluster로 묶인 노드의 집합은 하나의 큰 컴퓨팅 자원처럼 사용될 수 있습니다. 
Kubernetes의 도입으로 특정 노드에 직접 접근할 필요가 없어졌고 하나의 노드에 작업이 집중되는 것을 막을 수 있었습니다.

#### Remark: Ray Cluster

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/10.png" alt="Kubernetes" width="60%">
  <figcaption style="text-align: center;"> [그림10] - Ray Cluster [10]</figcaption>
</p>
</figure>

Regression Test를 하는데 너무 많은 시간이 소요된다면 업무의 병목이 될 수 있습니다.
Ray Cluster는 병렬로 실험을 진행하기 때문에 Regression Test에 걸리는 시간을 줄일 수 있습니다.

Ray Cluster는 헤드 노드와 워커 노드로 구성됩니다. 
헤드 노드에서는 작업들을 워커 노드에 분배하여 병렬로 테스크를 수행합니다. 
Ray Cluster는 작업들을 병렬적으로 처리하여 Regression Test를 빠르게 진행할 수 있습니다.
Ray Autoscaler는 Cluster의 자원상황을 고려하여 워커 노드의 개수를 동적으로 조절할 수 있습니다 [[3]](#ref-3).

MRX-Hosted Runner의 역할은 특정 Device내에서 Container로 Regression Test를 진행하는 것이 아닙니다. 
미리 정의된 컴퓨팅 자원 스펙에 해당하는 Ray Cluster를 만드는 것입니다 [[3]](#ref-2). 
여기서 Ray Cluster의 역할은 Regression Test를 병렬적으로 진행하기 위한 목적으로 사용되고 작업이 끝나게 되면 Ray Cluster는 사라지게 됩니다. 

> 참고로 [그림10]에서 구성한 Cluster와 Ray Cluster는 다른 역할을 합니다. 

Reviewer 3: 그림10의 제목이 Ray Cluster인데
이 그림이 Ray Cluster와 다른 역할을 한다고 기술되어 있어 혼란스럽습니다.

Author: 네 반영했습니다.
그림이 추가되고 제거되는 과정속에서 그림 번호가 얼라인이 안맞았습니다.

[그림10]은 자원자체를 묶는 작업을 의미한다면 Ray Cluster는 이미 묶인 자원을 활용하는 것입니다. 

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/11.png"  width="60%">
  <figcaption style="text-align: center;">[그림11] - Pipeline #4 </figcaption>
</p>
</figure>

이제 Kubernetes 그리고 Ray Cluster를 활용하여 [그림11]과 같은 Pipeline을 구축하였습니다. 
Repository에 의존성을 제거하였으며 Docker Image도 미리 만들어둔 Image를 활용하였습니다. 
또한 Device에 대한 의존성을 제거하여 내부의 컴퓨팅 자원을 더욱 효율적으로 사용할 수 있었습니다.

## Are You Sure? Yes!

GitHub Action에서 Trigger Event Type에 대해서 정할 수 있습니다.

> 여러 논의 끝에 Pull Request가 Approve될 때와 선택적으로 테스트를 진행하였습니다.

Reviewer 3: ```suggestion
여러 논의 끝에 Pull Request의 리뷰가 완료되었을 때 테스트가 수행되거나 또는 필요시 테스트 수행을 요청할 수 있도록 설정하였습니다.
```

이를 위해서 GitHub Actions의 Pull Request Review와 Workflow Dispatch Event를 사용하였습니다.
또한 Regression Test가 최신 브랜치기준으로 실행하는 것을 강제하기 위해서 Require branches to be up to date before merging 옵션을 선택하였습니다.

개발자들이 코드를 병합하는 과정은 다음과 같이 변경되었습니다.

1. Pull-Request를 통해 작업내용을 푸쉬합니다.
2. 유닛테스트로 코드를 검증합니다.
3. **(Optional: Workflow Dispatch) PR내용에 따라 Regression Test로 성능을 검증합니다.**
4. 작업내용을 동료들이 리뷰합니다.
5. **(Automatically: Pull Request Review) Regression Test로 성능을 검증합니다.**
6. 코드를 병합니다.

#### Remark: GitHub Actions Event Type

GitHub Actions는 특정 이벤트에 대하여 정해진 테스크를 수행할 수 있습니다.
예를 들어 Pull Request 이벤트가 발생할 때 테스크를 수행하고 싶다면 아래와 같이 작성하면 됩니다.

```yml
name: Regression_Test
on:
  pull_request:
```
매 Pull Request 혹은 Push마다 Regression Test를 수행한다면 너무 많은 실험을 진행해야합니다.

> 이런 문제를 해결하기 위해서 새로운 유형의 트리거 이벤트가 필요했습니다. 

Reviewer 3: 너무 많은 실험이 진행될 때 발생하는 문제가 무엇인가요? 중간이 빠진 느낌입니다.

Author: 반영했습니다.

Workflow Dispatch는 수동으로 GitHub Action을 수행할 수 있으며 [그림12]와 같이 GitHub UI를 통해서 쉽게 실행할 수 있습니다.
매 Pull Request 혹은 Push마다 Regression Test를 수행한다면 너무 많은 실험을 진행해야합니다.
코드리뷰가 끝난 후에 Regression Test를 수행하기 위하여 Workflow Dispatch를 선택하였습니다.

이제 마우스 클릭으로 GitHub Web에서 Regression Test를 실행할 수 있습니다.

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/12.gif"  width="60%">
  <figcaption style="text-align: center;">[그림12] - Click for Regression Test </figcaption>
</p>
</figure>

또한 Pull Request에 대한 Approve 이벤트 발생시 자동으로 실행시켜주기 위해 Pull Request Review 이벤트를 사용했습니다.

Workflow Dispatch와 Pull Request Review 트리거 이벤트를 사용하고 싶다면, 다음과 같이 작성하면 됩니다.

```yml
on:
  pull_request_review:
    types: [submitted]
  workflow_dispatch:

jobs:
  Regression_Test:
    if: (github.event.review.state == 'approved' || github.event_name == 'workflow_dispatch')
```

Regression Test Pipeline의 모습을 [그림13]으로 도식화했습니다.
GitHub에서 미리 설정한 Event Type에 해당하는 Event가 발생하면 MRX-Hosted-Runner에게 Regression Test를 요청합니다.
MRX-Hosted-Runner는 Ray Cluster를 구성합니다.
학습 및 실험을 진행할 때는 MLflow(중앙화된 실험기록 서비스)에 실험정보를 로깅하고 학습이 끝나면 이에 대한 정보를 GitHub에 전달합니다 [[7]](#ref-7).

> 현재는 해당 PR에 Comment로 MLflow 실험링크를 달아주는 방식으로 사용중입니다.

Reviewer 3: MLFlow을 본 포스팅에서 꼭 언급할 필요가 있을까요?

Author: 네 필수적이지 않아보입니다. 제거하겠습니다.

이를 통해서 팀 전체적으로 실험결과를 쉽게 확인할 수 있습니다.

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/13.png"  width="80%">
  <figcaption style="text-align: center;">[그림13] - Pipeline Overview </figcaption>
</p>
</figure>

#### Remark: GitHub Branch Protection Rule

GitHub에서는 특정 브랜치에 병합하기 위해서 필요한 상태확인(Status Check)을 지정할 수 있습니다.
Regression Test가 진행되지 않은 경우 병합을 못하도록 설정할 수 있습니다.

이는 다음과 같은 과정을 통해 설정할 수 있습니다.

1. Repository에 Settings을 클릭합니다.
2. 왼쪽 사이드바에 Branches를 클릭합니다.
3. Branch protection rules아래의 add rule을 클릭합니다.
4. Protect matching branches아래의 Require status checks to pass before merging을 클릭합니다.
5. Require branches to be up to date before merging을 클릭합니다.

> 6. Regression Test를 선택합니다.

Reviewer 3: 그림을 추가하면 어떨까요?

Author: ![image](https://user-images.githubusercontent.com/27891090/113529529-9d67f600-95fe-11eb-98a3-1bdb82a922a0.png)
추가해봤는데 어떤가요?

Regression Test에서 정상작동한 브랜치에 대해서 **Are You Sure?** 라고 누가 묻는다면 이제는 자신있게 **Yes!**라고 할 수 있습니다.

> 이런 변화는 PR에 대한 자신감을 키워주고 결과적으로 견고하면서 빠른 협업이 가능합니다.

Reviewer 3: ```suggestion
이런 변화는 코드변경에 대한 자신감을 키워주었고 나아가 견고하면서 빠른 협업을 가능하게 했습니다.
```

## 마치며

이번 포스팅에서는 Machine Learning Software의 Regression Test에 대해서 다뤘습니다. 

개발을 하다보면 의도치 않은 부작용들이 발생하게 되며 이런 부작용은 발견하기 어렵습니다. 
이런 부채들이 쌓인 후 디버깅하게 되면 생각보다 훨씬 큰 비용을 치뤄야 합니다. 

> 이런 문제를 보다 효과적으로 대처하기 위해서 Regression Test를 제안하게 되었습니다.

Reviewer 3: 문제가 무엇인지에 대해 한 문장으로나마 명확히 정리해주시면 어떨까요?
결론에 `이런 문제` (366라인) / `비슷한 문제` (376라인) 와 같이 문제가 무엇인지 명시적으로 설명하고 있지 않아 포스팅의 결론이 다소 깔끔하게 정리되지 않는 것 같습니다.

문제상황 -> 아이디어 -> 개선 -> 기대효과

Author: 네 반영하겠습니다.

Regression Test Pipeline을 구성하기 위해서 여러가지 추상화과정이 필요했습니다. 
Repository에 독립적으로 작동할 수 있어야 했습니다. 
또한 Machine Learning Software(AI)는 많은 컴퓨팅 자원을 요구하기 때문에 효율적인 자원사용이 필요했습니다. 
이를 위해서 Kubernetes를 활용하여 컴퓨팅 자원을 가상화하였습니다. 

> Regression Test를 통해서 문제를 빠르게 발견할 수 있었습니다.
> 그리고 클릭 한 번으로 실험을 진행할 수 있다는 것도 굉장히 매력적인 일이였습니다.

Reviewer 3: ```suggestion
Regression Test Pipeline의 도입을 통해 코드변경에서 발생하는 문제를 빠르게 발견할 수 있었습니다.
또한 특정시점에 실험을 자동으로 수행하하거나 클릭 한 번으로 실험을 진행할 수 있다는 것도 굉장히 매력적인 일이였습니다.
```

---
layout: post
title: Regression Test, Are you sure?
author: wontak ryu
categories: [test]
image: assets/images/2020-02-10-Regression-Test/thumbnail.gif
---

마키나락스는 AI 기술개발을 넘어서 AI 제품화과정으로 나아가고 있습니다. 
제품화과정으로 나아가는 여정 속에서 재미있는 엔지니어링 이슈들이 생겼습니다.
제품의 실체화에 가까워질수록 시장에서 발생하는 요구사항을 지속적으로 빠르게 반영해야 합니다.
이를 위해 Machine Learning Software의 신속하고 정확한 성능 검증을 필요로 하게 되었습니다.
유닛테스트 검증만으로 모델의 학습성능에 대한 검증이 어려우므로, 성능검증에 대한 다른 방법이 필요했습니다.

이번 포스트에서는 Machine Learning Software에 대한 성능검증을 어떤 방식으로 진행하고 있는지 공유드리겠습니다.

## Problem: Can't find the cause of the lower performance!

우선 AI Product를 개발하며 겪었던 문제에 대해서 공유드리겠습니다. 

개발자들은 작업을 완료하면 다음의 과정을 통해 코드를 병합(Merge)합니다.

1. Pull-Request를 통해 작업내용을 푸쉬합니다.
2. 유닛테스트로 코드를 검증합니다.
3. 작업내용을 동료들이 리뷰합니다.
4. 코드를 병합니다.

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/1.png"  width="60%">
  <figcaption style="text-align: center;">[그림1] - Gitflow Workflow [1]</figcaption>
</p>
</figure>

하지만 유닛테스트 만으로는 모델의 성능저하를 일으키는 변경사항을 알아챌 수 없다는 문제가 있었습니다. 

> 예를 들어, 활성함수(activation function)로 ReLU를 사용했을때 특정 모델의 성능저하가 일어난 사건이 있었습니다. 

Reviewer 2: ```suggestion
예를 들어, 활성함수(Activation Function)로 ReLU를 사용했을때 특정 모델의 성능저하가 일어난 사건이 있었습니다. 
```

Reviewer 2: 사용했을 때 보다, 변경했을 때가 흐름상 좋을 것 같습니다
```suggestion
예를 들어, Leaky ReLU로 사용하던 활성함수(activation function)를 ReLU로 변경했을때 특정 모델의 성능저하가 일어난 사건이 있었습니다. 
```

Author: 예를 들어, Leaky ReLU로 사용하던 활성함수(Activation Function)를 ReLU로 변경했을때 치명적인 성능저하가 일어난 사건이 있었습니다. 

이런식으로 변경하였습니다.

```yml
name: Regression_Test

jobs:
  regression_test:
    runs-on: [self-hosted, ubuntu-18.04]
```

### Pipeline #1: Dependent on Repository

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/5.png"  width="60%">
  <figcaption style="text-align: center;">[그림5] - Pipeline #1</figcaption>
</p>
</figure>

첫 번째로 구현한 Pipeline은 [그림5]에서 볼 수 있습니다. 
MRX-Hosted Runner가 Regression Test 대상이 되는 Repository의 Requirements(필요환경)를 미리 가지고 있습니다. 
학습에 필요한 데이터의 경우 원격 저장소에 저장해두고 요청 시 접근하여 사용합니다. 
GitHub에서 테스트요청을 보내면 Regression Test를 진행하게 됩니다. 

이런 구조는 한 Repository에 의존성을 가지게 된다는 문제를 가지고 있습니다.
특정 Repository를 위한 MRX-Hosted Runner가 다른 Repository를 운영할 수 없습니다.

### Pipeline #2: Independent on Repository, But Inefficient

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/6.png"  width="60%">
  <figcaption style="text-align: center;">[그림6] - Pipeline #2</figcaption>
</p>
</figure>

두 번째로 구현한 Pipeline은 [그림6]에서 볼 수 있습니다. 
Pipeline #1과 다르게 MRX-Hosted Runner가 Repository에 정의된 Dockerfile을 기반으로 Regression Test Container를 만듭니다. 
이를 통해서 Repository에 의존성을 가지던 문제를 해결할 수 있었습니다. 

> 하지만 Docker Image를 Build하는 작업은 상당히 오랜시간이 걸리기 때문에 비효율적이라는 문제가 있었습니다.

Reviewer 2: (잘 모르고 드리는 말씀입니다) 단순히 Docker Image를 빌드하는 시간이 오래 걸리나요? ML 실행 환경을 구성하는데 오래걸릴 것 같습니다(ex, pytorch). 맞다면 왜 오래걸리는지 조금 더 구체적으로 적으면 좋을 것 같습니다.

Author: ```suggestion
하지만 Docker Image를 Build하는 작업은 상당히 오랜시간이 걸리며 필요한 라이브러리가 변하지 않을때도 중복적으로 Docker Image를 빌드해야합니다.
Machine Learning Software의 경우 사용하는 라이브러리의 용량이 큰 것을 고려해보면 이는 상당히 비효율적입니다.
```
이런식으로 수정했습니다.

### Pipeline #3: InDependent on Repository, But!

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/7.png"  width="60%">
  <figcaption style="text-align: center;">[그림7] - Pipeline #3</figcaption>
</p>
</figure>

> 첫 번째로 구현한 Pipeline은 [그림7]에서 볼 수 있습니다. 

Reviewer 2: ```suggestion
세 번째로 구현한 Pipeline은 [그림7]에서 볼 수 있습니다. 
```

Docker Image는 Requirements가 변경되었을 때만 Update가 필요했습니다. 

> 따라서 미리 DockerImage를 만들어 두고 MRX-Hosted Runner가 이를 받아서 사용하도록 구조를 변경하였습니다. 

Reviewer 2: 문장이 잘 이해되지 않습니다. 둘다 미리 Docker Image를 만드는 부분은 동일 한것 같은데 어떤 차이가 있나요? 

DockerImage 와 Docker Image  통일하면 좋을 것 같습니다

Author: ```suggestion
Pipeline #2의 경우 테스트 요청마다 Docker Image를 빌드해야합니다.
반면에, Pipeline #3는 테스트 요청마다 미리 만들어둔 Docker Image를 Pull하여 사용합니다.
```

Author: 위와 같은 내용을 추가하였습니다.

헤드 노드에서는 작업들을 워커 노드에 분배하여 병렬로 테스크를 수행합니다. 
Ray Cluster는 작업들을 병렬적으로 처리하여 Regression Test를 빠르게 진행할 수 있습니다.
Ray Autoscaler는 Cluster의 자원상황을 고려하여 워커 노드의 개수를 동적으로 조절할 수 있습니다 [[3]](#ref-3).

MRX-Hosted Runner의 역할은 특정 Device내에서 Container로 Regression Test를 진행하는 것이 아닙니다. 
미리 정의된 컴퓨팅 자원 스펙에 해당하는 Ray Cluster를 만드는 것입니다 [[3]](#ref-2). 
여기서 Ray Cluster의 역할은 Regression Test를 병렬적으로 진행하기 위한 목적으로 사용되고 작업이 끝나게 되면 Ray Cluster는 사라지게 됩니다. 
참고로 [그림9]에서 구성한 Cluster와 Ray Cluster는 다른 역할을 합니다. 
[그림9]은 자원자체를 묶는 작업을 의미한다면 Ray Cluster는 이미 묶인 자원을 활용하는 것입니다. 

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2020-02-10-Regression-Test/12.png"  width="60%">
  <figcaption style="text-align: center;">[그림12] - Pipeline #4 </figcaption>
</p>
</figure>

이제 Kubernetes 그리고 Ray Cluster를 활용하여 [그림12]과 같은 Pipeline을 구축하였습니다. 
Repository에 의존성을 제거하였으며 Docker Image도 미리 만들어둔 Image를 활용하였습니다. 
또한 Device에 대한 의존성을 제거하여 내부의 컴퓨팅 자원을 더욱 효율적으로 사용할 수 있었습니다.

## Are You Sure? Yes!

GitHub Action에서 Trigger Event Type에 대해서 정할 수 있습니다.
여러 논의 끝에 Pull Request의 리뷰가 완료되었을 때 테스트가 수행되거나 또는 필요시 테스트 수행을 요청할 수 있도록 설정하였습니다.
이를 위해서 GitHub Actions의 Pull Request Review와 Workflow Dispatch Event를 사용하였습니다.

> 또한 Regression Test가 최신 브랜치기준으로 실행하는 것을 강제하기 위해서 Require branches to be up to date before merging 옵션을 선택하였습니다.

Reviewer 2: 옵션 1이므로 quotes를 사용하면 좋을 것 같습니다
```suggestion
또한 Regression Test가 최신 브랜치기준으로 실행하는 것을 강제하기 위해서 'Require branches to be up to date before merging' 옵션을 선택하였습니다.
```
